lib/moab/storage_services.rb

```diff
@@ -14,80 +14,79 @@ module Moab
   #   All rights reserved.  See {file:LICENSE.rdoc} for details.
   class StorageServices
     # @return [StorageRepository] an instance of the interface to SDR storage
-    @@repository = Moab::StorageRepository.new
 
     def self.repository
-      @@repository
+      @repository ||= Stanford::StorageRepository.new
```

Comment:
So... I haven't done a whole new web search to refresh myself on the problem, but when I was looking into Ruby multi-threading issues and Sidekiq/Resque stuff a couple years ago, I found this blog post (one of the things I actually bookmarked on the subject) particularly helpful: https://bearmetal.eu/theden/how-do-i-know-whether-my-rails-app-is-thread-safe-or-not/

Anyway, one of my takeaways from it was that class vars (`@@`) and class instance vars (`@` ivars in a class method) are equally dangerous, and similarly subject to multi-threading issues.

Re-reading that article today, I have a few thoughts:
1. The most dangerous pitfalls in this sort of multi-threading situation seem to be around mutations where a calculation might overwrite shared state in such a way that a stale value is used in a calculation that then overwrites a fresher value.  The simple but common pitfall they point out is something like a shared class var counter incremented by the `+=` operator, which is actually two operations... but as they observe, so is the `||=`.  But in the case of our code here, the possible overwrite will almost certainly just overwrite the existing value with the same value, because the `Stanford::StorageRepository` instance will be based on the same configuration each time (at least the way we use moab-versioning -- maybe this won't be the case for other consumers? :shrug:).  Which is to say, maybe this is technically still not thread safe, but also, maybe it doesn't matter?  But this does worry me a bit in terms of finding other thread un-safe operations in the gem, because as that post observed, a class ivar is hard to distinguish from an object ivar, at least just by grepping for `@`.
2. One way out: we could just not memoize this value.  I suspect that we wouldn't experience a noticeable performance hit, because other perservation versioning operations like updating files and checking checksums are relatively way more expensive... but I'm not sure the performance hit would be negligible.  "Premature optimization is the root of all evil" -some famous computer scientist (I think it was Wirth, but am not sure off-hand 😂)
3. Another way out: per that blog post, we could memoize in a thread local var, e.g. `Thread.current[:repository] ||= Stanford::StorageRepository.new`.  That might be my preference: seems like it'd be the least disruptive change in terms of the expected behavior and interface from the consumer's perspective, while also providing the desired thread safety.  As the blog post points out, it sort of smells like a global var, but I don't think it smells anymore like that than what the gem does at present 🙂 
4. I'm not totally sure, but I don't think explicitly splitting the `||=` into separate atomic operations would make a difference (e.g. a `return` guard clause if the class var exists, and memoization if the guard clause falls through).
5. We could make it the responsibility of the consumer to instantiate `Stanford::StorageRepository` locally, but that sounds like a pain to coordinate across the codebases, so I'd love to avoid that approach.

This all makes me sort of want to just grep for memoization in this gem to look for other similar issues... a super cursory `git grep '||='` makes me think that the other memoizations are actually just regular object ivars (as opposed to class instance vars like this).  But not 100% sure about that.

Reply:
TLDR i vote that we tweak this line to be `Thread.current[:repository] ||= Stanford::StorageRepository.new`, and see how that works.

Reply:
one last thing, not really a suggestion, but interested in others' thoughts on it: one thing that the blog post suggests is a `Mutex#synchronize` wrapper around the thread-unsafe code... but the blog post also suggests creating and storing the `Mutex` in a way that seems not thread-safe, because it suggests putting it in a `@@` class var?  and i'm having a hard time thinking of how one would even share the mutex in a safe way, given that.

Reply:
nope, wrong attribution, it was knuth (of course)

Reply:
ugh, an now, after thinking about this more and reading about it more, i'm left wondering if the current implementation on `main` is maybe thread-safe afterall... if the class is only instantiated once per process, and the class var is only set at instantiation and never mutated, is there actually a worry around concurrency and atomicity?

maybe we should flag this for post-standup discussion monday.  might help to have multiple sets of eyes on this, esp if some of the viewers better grok the way ruby classes are managed by the interpreter than i do.

Reply:
Post standup discussion (or infra dev) is totally warranted here I think. Thanks for the thoughtful comments :-)

Reply:
np!